File: cpp/src/0383.cpp
/*
[383. Ransom Note (Easy)](https://leetcode.com/problems/ransom-note/)

@topic: Hash Table
@topic: String
@topic: Counting

Constraints:
  - `1 <= ransomNote.length, magazine.length <= 10^5`
  - `ransomNote` and `magazine` consist of lowercase English letters.
*/

#include <string>

using std::string;

class Solution {
   public:
    bool canConstruct(string ransomNote, string magazine) {
        if (ransomNote.size() > magazine.size()) return false;
        int freq[26]{};
        for (char c : magazine) ++freq[c - 'a'];
        for (char c : ransomNote) if (--freq[c - 'a'] < 0) return false;
        return true;
    }
};

#include <cassert>
#include <chrono>
#include <iostream>
#include <vector>

using std::tuple;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::microseconds;
using std::chrono::system_clock;

int main() {
    Solution o;

    vector<tuple<string, string, bool>> CASES = {
        {"a", "b", false},
        {"aa", "ab", false},
        {"aa", "aab", true},
    };

    for (auto& [ransomNote, magazine, excepted] : CASES) {
        assert(o.canConstruct(ransomNote, magazine) == excepted);
    }

    auto start = system_clock::now();
    for (auto& [ransomNote, magazine, _] : CASES) {
        for (int i = 0; i < 100000; ++i) {
            o.canConstruct(ransomNote, magazine);
        }
    }
    auto end = system_clock::now();
    std::cout << duration_cast<microseconds>(end - start).count() << std::endl;
}

File: python/src/0383.py
from collections import Counter


class Solution:

    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        return len(ransomNote) <= len(magazine) and not Counter(ransomNote) - Counter(magazine)


if __name__ == '__main__':
    import unittest
    from timeit import timeit

    CASES = (
        ('a', 'b', False),
        ('aa', 'ab', False),
        ('aa', 'aab', True),
        ('aab', 'aab', True),
    )

    class Test(unittest.TestCase):

        def test(self):
            canConstruct = Solution().canConstruct
            for ransomNote, magazine, excepted in CASES:
                actual = canConstruct(ransomNote, magazine)
                self.assertEqual(actual, excepted)

    unittest.main()

    canConstruct = Solution().canConstruct
    for ransomNote, magazine, excepted in CASES:
        print(timeit(lambda: canConstruct(ransomNote, magazine), number=100000))
